### ISSUE TITLE
Mysterious error in ssubst

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#799]
Imported from SourceForge on 2024-07-03 11:29:55
Created by **[npstrick](https://sourceforge.net/u/npstrick/)** on 2021-01-10 08:48:59
Original: https://sourceforge.net/p/maxima/bugs/3700

---

I am using Maxima 5.43.2 with GCL 2.6.12 under Ubuntu 20.04.1 LTS.  (This is what `apt` gives me as the latest version of maxima.  I have not tried any other kind of installation.)

For certain long strings `s` containing many backslashes, `ssubst("\\\\","\\",s)` gives 
```
`ssubst': unsuitable start or end position.
 -- an error.
```
More precisely, I start Maxima, then enter
```
s : <long thing as specified below>$
ssubst("\\\\","\\",s);
ssubst("\\\\","\\",s);
```
The first `ssubst` gives an error as described.  The second (identical) `ssubst` gives an apparently correct  answer, with all the backslashes in `s` doubled up.  The string `s` (which I have tried to minimise) is as follows:
```
"[0,\"\",\"\\\\(x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\x\\\\) \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\)  \\\\(x\\\\)\",[[[A,false,\"\\\\(x\\\\)\",\"\"],[A,false,\"\\\\(x\\\\) \\\\(\\\\x\\\\)\",\"\\\\(\\\\x\\\\x\\\\)\\\\(\\\\x\\\\x\\\\x\\\\) \\\\(\\\\x\\\\x\\\\x\\\\x\\\\x\\\\x\\\\) \\\\(\\\\x\\\\)  \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\x\\\\) \\\\(\\\\x\\\\)  \\\\(x\\\\)  \\\\(x\\\\x\\\\) \\\\(\\\\x\\\\) \\\\(x\\\\)\"],[A,true,\"\\\\(x\\\\)\",\"\"],[C,false,\"\\\\(x\\\\)  \\\\(x\\\\x\\\\x\\\\x\\\\)\",\"\\\\(x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\x\\\\) \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\)  \\\\(x\\\\)\"]],[[C,false,\"\\\\(x\\\\)  \\\\(x\\\\x\\\\x\\\\x\\\\)\",\"\\\\(x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\x\\\\) \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\)  \\\\(x\\\\)\"]],[[A,false,\"\\\\(x\\\\)\",\"\"],[A,false,\"\\\\(x\\\\) \\\\(\\\\x\\\\)\",\"\\\\(\\\\x\\\\x\\\\)\\\\(\\\\x\\\\x\\\\x\\\\) \\\\(\\\\x\\\\x\\\\x\\\\x\\\\x\\\\x\\\\) \\\\(\\\\x\\\\)  \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\x\\\\) \\\\(\\\\x\\\\)  \\\\(x\\\\)  \\\\(x\\\\x\\\\) \\\\(\\\\x\\\\) \\\\(x\\\\)\"],[A,true,\"\\\\(x\\\\)\",\"\"]],[[C,false,\"\\\\(x\\\\)  \\\\(x\\\\x\\\\x\\\\x\\\\)\",\"\\\\(x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\x\\\\) \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\)  \\\\(x\\\\)\"]],[C,false,\"\\\\(x\\\\)  \\\\(x\\\\x\\\\x\\\\x\\\\)\",\"\\\\(x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\x\\\\) \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\)  \\\\(x\\\\)\"],\"\\\\(x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\x\\\\) \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\)  \\\\(x\\\\)\",\"\\\\(x\\\\)  \\\\(x\\\\x\\\\x\\\\x\\\\)\\\\(x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\x\\\\) \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\x\\\\)  \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\)  \\\\(x\\\\) \\\\(x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\x\\\\) \\\\(x\\\\x\\\\x\\\\x\\\\)  \\\\(x\\\\x\\\\x\\\\x\\\\x\\\\) \\\\(x\\\\)  \\\\(x\\\""
```

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 11:29:57
Created by **[npstrick](https://sourceforge.net/u/npstrick/)** on 2021-01-10 11:48:17
Original: https://sourceforge.net/p/maxima/bugs/3700/#139b

---

Actually, here is a much simpler case.  If we set `s` to be a string consisting of the letter "a" repeated  1355 (or more) times, and evaluate `ssubst("b","a",s)`, then we get the same error.  So in particular, special properties of the backslash character do not seem to be relevant.

I note that `ssubst` is defined at line 1129 of `/usr/share/maxima/5.43.2/share/stringproc/stringproc.lisp`, and that the definition is recursive, with one nested function invocation for each occurrence of the string to be substituted.  I am not familiar with the internals of lisp.  Is this really executed recursively, or is that optimised away somehow?  Is the depth of the recursion likely to be a problem?

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 11:30:04
Created by **[npstrick](https://sourceforge.net/u/npstrick/)** on 2021-01-10 15:09:05
Original: https://sourceforge.net/p/maxima/bugs/3700/#1614

---

Output of `build_info()` is as follows:
```
Maxima version: "5.43.2"
Maxima build date: "2020-02-21 05:22:38"
Host type: "x86_64-pc-linux-gnu"
Lisp implementation type: "GNU Common Lisp (GCL)"
Lisp implementation version: "GCL 2.6.12"
User dir: "/root/.maxima"
Temp dir: "/tmp"
Object dir: "/root/.maxima/binary/5_43_2/gcl/GCL_2_6_12"
Frontend: false
```

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 11:30:08
Created by **[peterpall](https://sourceforge.net/u/peterpall/)** on 2021-01-10 21:25:43
Original: https://sourceforge.net/p/maxima/bugs/3700/#1614/92c2

---

I personally use the maxima version from https://code.launchpad.net/~maxima-developers/+archive/ubuntu/maxima-nightly, that was compiled using sbcl. It is both newer than your maxima version *and* has been compiled using sbcl, not gcl. A string consisting of 10000 "a" is replaced to 10000 "b" characters just fine, for me.

=> The problem is either caused by gcl or by using the old maxima version.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 11:30:15
Created by **[l_butler](https://sourceforge.net/u/l_butler/)** on 2021-01-12 01:59:29
Original: https://sourceforge.net/p/maxima/bugs/3700/#139b/c09f

---

Neil, thanks for tracking this down.

Attached is a patched stringproc.lisp (and a diff of the changes).
I have eliminated the recursion in ssubst and replaced it with a while
loop, and I have introduced a minor optimization is ssubstfirst to
prevent a second search.

Can you confirm this works for you?

Thanks,
Leo





Attachments:
- https://sourceforge.net/p/maxima/bugs/_discuss/thread/72bc97e135/139b/c09f/attachment/stringproc.lisp
- https://sourceforge.net/p/maxima/bugs/_discuss/thread/72bc97e135/139b/c09f/attachment/stringproc.lisp.diff

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 11:30:19
Created by **[npstrick](https://sourceforge.net/u/npstrick/)** on 2021-01-12 10:04:33
Original: https://sourceforge.net/p/maxima/bugs/3700/#139b/c09f/cb3a

---

Thanks.  I just replaced the file `/usr/share/maxima/5.43.2/share/stringproc/stringproc.lisp` with the one that you sent me; I am not sure if there is supposed to be a compiled version anywhere.   This does indeed eliminate the bug for me.  It is not very fast, though; 3 or 4 seconds for a string of length 16384.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 11:30:26
Created by **[l_butler](https://sourceforge.net/u/l_butler/)** on 2021-01-12 19:17:54
Original: https://sourceforge.net/p/maxima/bugs/3700/#21ac

---

Fixed in [commit 6f40fdce7](https://sourceforge.net/p/maxima/code/ci/6f40fdce72e40eb3d0af8ce600dca0433e7b58ef/).

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 11:30:30
Created by **[l_butler](https://sourceforge.net/u/l_butler/)** on 2021-01-12 19:18:41
Original: https://sourceforge.net/p/maxima/bugs/3700/#4fa5

---

- **labels**:  --> stringproc, string, ssubst
- **status**: open --> closed
